feat: add MFA (WebAuthn) docs (#30049)
supabase/supabase: apps/docs/content/guides/auth/auth-mfa/webauthn.mdx

```

As with enrollment, developers have the option of performing verification over multiple steps.

```
const { data: factors, error } = await supabase.auth.mfa.listFactors()
const { data, error } = await supabase.auth.mfa.challenge({
	factorId: factors.find((type) => type === 'webauthn'), // use first verified factor
	})

const credential = await generateAuthenticationOptions({authenticator: {userVerification, timeout, authenticatorAttachment}})

const { data: factor, error } = await supabase.auth.mfa.verify({ factorId, credential, useMultiStep: true})
```

Note that you do not have to specify a `factorId` parameter. Unlike MFA (Phone) and MFA (TOTP), authentication via the WebAuthn factor does not depend on the specific factor in use. Auth will check the assertion against all verified factors owned by the user, as well as the existing WebAuthn session, and make a decision on authentication based on the information given.

## Factor Management

## Best Practices

You can use the following directives to help control availability of a credential in an iframe. Add these directives so that WebAuthn pop-ups for authentication can be shown in the iframe.

> - [`publickey-credentials-create`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Permissions-Policy/publickey-credentials-create): Controls the availability of [`navigator.credentials.create()`](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/create) with the `publicKey` option.

Author: Might not be needed

---

## What is WebAuthn?

WebAuthn (Web Authentication) is a web standard that enables secure, passwordless authentication using public-key cryptography. It allows users to log in to websites and applications using built-in authenticators like fingerprint sensors, facial recognition, or external devices like security keys.

### When to Use WebAuthn?

Consider implementing WebAuthn when:

1. You want to enhance security beyond traditional passwords
2. You aim to provide a more user-friendly authentication experience
3. You need to reduce costs associated with SMS-based authentication
4. You want to offer a hardware backup to existing authentication factors

WebAuthn is particularly suitable for security-conscious applications that require a user-friendly, passwordless solution.

### Pricing and Availability

WebAuthn MFA is part of the Auth Advanced MFA Add-on. Here's the breakdown of pricing:

- Cost: $75 per month for the first project in the organization
- Additional projects: $10 per month each
- Included in the Add-on: Phone and other advanced MFA features

## Types of Authenticators

> An authenticator is a component that securely generates and stores cryptographic keys associated with a particular relying \***\*party (web page origin) and a credential \*\***ID (a unique identifier for the key). The authenticator uses these keys to authenticate the user to the relying party without directly storing user-specific data like the user ID.

Reviewer: formatting looks off in the preview 
<img width="973" alt="image" src="https://github.com/user-attachments/assets/e9a9614b-7e21-40e9-b798-2f26cce7d727">

### Add enrollment flow

An enrollment flow provides a UI for users to set up additional authentication factors. Most applications add the enrollment flow in two places within their app:

1. Right after login or sign up.
   This allows users quickly set up Multi Factor Authentication (MFA) post login or account creation. Where possible, encourage all users to set up MFA. Many applications offer this as an opt-in step in an
   effort to reduce onboarding friction.
2. From within a settings page.
   Allows users to set up, disable or modify their MFA settings.

As far as possible, maintain a generic flow that you can reuse in both cases with minor modifications.

You can enroll a WebAuthn factor with a single step:

> ```

Reviewer: for the syntax highlighting
```suggestion
```ts
```

import { browserSupportsWebAuthn } from '@supabase/supabase-js'
if (supabase.auth.browserSupportsWebAuthn()) {
    const { data, error } = await supabase.auth.mfa.enroll({ factorType: 'webauthn'})
}) else {
   // Use another MFA method, such as MFA (Phone)
   // const { data, error } = await supabase.auth.mfa.enroll({ factorType: 'phone', phone: '<your-phone-number>' })
};
```

Or in multiple-steps by setting the `useMultiStepEnroll` parameter to true.

> ```

Reviewer: ```suggestion
```ts
```

const { data: { factorId }, error } = await supabase.auth.mfa.enroll({ factorType: 'webauthn', useMultiStep: true })

const { data: {options}, error } = await supabase.auth.mfa.challenge({ factorId })
const options = await supabase.auth.generateRegistrationOptions{
		options,
		rp_name,
		user: {
			name,
			display_name,
		}
		authenticator_options: {
			attestation: 'direct', // or developer choice of 'enterprise'
			authenticatorAttachment,
			requireResidentKey,
			userVerification,
		}
	}
})

const credential = await navigator.credentials.create(data)
const { data, error } = await supabase.auth.mfa.verify({ factorId, credential })
```

> `useMultiStep` informs Auth to abstract away the complexities of interacting with the browser API. Multi Step enrollment is recommended only for Advanced Users who require a heightened level of customization in their application

Reviewer: ```suggestion
Setting `useMultiStepEnroll` to false informs Auth to abstract away the complexities of interacting with the browser API. Multi Step enrollment is recommended only for Advanced Users who require a heightened level of customization in their application
```

When using multi-step enroll there are a substantial number of fields to deal with. The fields are highlighted below:

| Field Type       | Field Name                | Configurability                                                                                      | Description                                                           |
| ---------------- | ------------------------- | ---------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------- |
| Non-Configurable | `excludeCredentials`      | Generated by auth server                                                                             | Prevents enrollment of multiple credentials                           |
| Non-Configurable | `rp.id`                   | Generated by auth server                                                                             | Relying Party ID                                                      |

> | Non-Configurable | `challenge`               | Generated by auth server                                                                             | Unique challenge for the authentication attempt                       |

Reviewer: not sure if the non-configurable options are implemented / enforced at the client lib level but if we don't even show these options in the client lib / accept them in the Auth API, then maybe worth leaving it out

| Configurable     | `authenticatorAttachment` | Developer choice                                                                                     | Preferred authenticator type ('platform' or 'cross-platform')         |
| Configurable     | `requireResidentKey`      | Developer choice                                                                                     | Whether a resident key is required (true or false)                    |
| Configurable     | `userVerification`        | Developer choice                                                                                     | User verification preference ('required', 'preferred', 'discouraged') |
| Configurable     | `rp.name`                 | Developer choice                                                                                     | Human-readable name for the Relying Party                             |

While configurable fields can be set by developers, it's important to choose appropriate values based on your security requirements and user experience considerations.

> Visit the MDN Documentation for a full description of each field [here](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialCreationOptions)

Reviewer: ```suggestion
Refer to the [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialCreationOptions) for a full description of each field.
```

## Add Login Flow

Once a user has registered a factor, they can proceed to log in. In single-step mode this is accomplished with:

```jsx
const { data, error } = await supabase.auth.mfa.verify({ factorType: 'webauthn' })
```

As with enrollment, developers have the option of performing verification over multiple steps.

> ```

Reviewer: ```suggestion
```ts
```

### Add enrollment flow

An enrollment flow provides a UI for users to set up additional authentication factors. Most applications add the enrollment flow in two places within their app:

1. Right after login or sign up.
   This allows users quickly set up Multi Factor Authentication (MFA) post login or account creation. Where possible, encourage all users to set up MFA. Many applications offer this as an opt-in step in an
   effort to reduce on-boarding friction.
2. From within a settings page.
   Allows users to set up, disable or modify their MFA settings.

As far as possible, maintain a generic flow that you can reuse in both cases with minor modifications.

You can enroll a WebAuthn factor with a single step:

```ts

> import { browserSupportsWebAuthn } from '@supabase/supabase-js'

Reviewer 2: Does this need to be imported? Looks like it is a method on `supabase.auth` from the example below?

Author: Thanks! You're right the method `browserSupportsWebAuthn` shouldn't have a `supabase.auth` prefix

if (supabase.auth.browserSupportsWebAuthn()) {

>     const { data, error } = await supabase.auth.mfa.enroll({ factorType: 'webauthn'})

Author: ```suggestion
    const { data, error } = await supabase.auth.mfa.enroll({ factorType: 'webauthn' , friendlyName:'a_name'})
```

> if (supabase.auth.browserSupportsWebAuthn()) {
>     const { data, error } = await supabase.auth.mfa.enroll({ factorType: 'webauthn', friendlyName: 'my-friendly-name'})
> })

Reviewer 2: ```suggestion
if (supabase.auth.browserSupportsWebAuthn()) {
    const { data, error } = await supabase.auth.mfa.enroll({ factorType: 'webauthn', friendlyName: 'my-friendly-name'})
}
```

```

Or in multiple-steps by setting the `useMultiStep` parameter to true.

```ts
const {
  data: { factorId },
  error,
} = await supabase.auth.mfa.enroll({ factorType: 'webauthn', useMultiStep: true })

const {
  data: { options },
  error,
} = await supabase.auth.mfa.challenge({ factorId })

let challengeOptions = challengeData?.credential_creation_options.publicKey

// Modify challengeOptions as needed. See table below for full details.
// challengeOptions.attestation = 'indirect'

const credential = await supabase.auth.startAuthentication(challengeOptions)

const { data, error } = await supabase.auth.mfa.verify({ factorId, credential })
```

> Setting `useMultiStep` to false informs Auth to abstract away the complexities of interacting with the browser API. Multi Step enrollment is recommended only for Advanced Users who require more customization in their application

Reviewer 2: ```suggestion
Setting `useMultiStep` to false informs Auth to abstract away the complexities of interacting with the browser API. Multi Step enrollment is recommended only for Advanced Users who require more customization in their application.
```
